File: DeveloperTool/src/TfsExplorer/VSExtensions/Types/Optional.cs
using System;
using System.Collections;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace Spark.Infra.Types
{
    public class Optional<T> : IEnumerable<T>
    {
        private IEnumerable<T> Content { get; }
        private Optional(IEnumerable<T> content)
        {
            this.Content = content;
        }


        public static Optional<T> Some(T value)
        {
            return new Optional<T>(new T[] { value });
        }

        public static Optional<T> None()
        {
            return new Optional<T>(new T[0]);
        }

        public void Do(Action<T> action)
        {
            foreach(var item in this.Content)
            {
                action(item);
            }
        }

        public IEnumerator<T> GetEnumerator()
        {
            return this.Content.GetEnumerator();
        }

        IEnumerator IEnumerable.GetEnumerator()
        {
            return this.GetEnumerator();
        }
    }
}

File: Portal/WebClient/src/app/utils/optional.ts
export class Optional<T>{
   
    content : T[] = [];

    static some<Z>(item : Z) : Optional<Z>{
        var opt = new Optional<Z>();
        opt.content = [item];
        return opt;
    }

    static none<Z>() : Optional<Z>{
        return new Optional<Z>();
    }

    do(callBack: (item : T) => void){
        for(let i of this.content){
            callBack(i);
        }
    }

    any() : boolean{
        return this.content.length > 0;
    }
}